plugins/base/src/main/kotlin/renderers/html/HtmlCodeBlockRenderer.kt

```diff
@@ -0,0 +1,27 @@
+/*
+ * Copyright 2014-2023 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package org.jetbrains.dokka.base.renderers.html
+
+import kotlinx.html.FlowContent
+
+/**
+ * Provides an ability to override code blocks rendering differently dependent on the code language.
+ *
+ * Multiple renderers can be installed to support different languages in an independent way.
+ */
+public interface HtmlCodeBlockRenderer {
+
+    /**
+     * Whether this renderer supports given [language].
+     *
+     * [code] can be useful to determine applicability if [language] is not provided (empty string)
```

Comment:
I agree. Removed `code` from `isApplicable`.
But if we will think of [indented code blocks](https://spec.commonmark.org/0.30/#indented-code-blocks) - there is no possibility to define language there, and as so, now, there is no possibility to override rendering of such code blocks.

Reply:
Does it mean the auto-detection will also be removed from the mermaid plugin? https://github.com/whyoleg/dokka-mermaid/commit/ae19d44c2568e1b2f30a40c8e44c4bbeb226ff5e#diff-447dd6e1ceda7164c5b15fc93976a140abc6c9a28237837cbe2eb3bbfb0e0852L63-R50

I actually think it was useful to have `code` in `isApplicable` (either as a parameter or as an additional function):
1. Developers (or whoever writes markdown docs) are lazy and might just skip setting the language
2. It allows for two extensions for the same language, but for different types of code to be registered (for example, one can parse mermaid v1 and there other can parse mermaid v3, if they have differences)
3. Indented code blocks

Reply:
Yeah, if we will not provide `code` auto detection will not work.
To support all use cases we need to provide both arguments.
So `public fun isApplicable(language: String?, code: String): Boolean`
Also, as pointed below in other comment, better to make `language` nullable, to better understand why `code` provided at all (except documentation of course)

Reply:
@vmishenev WDYT? May be you have other ideas, on how we could support those cases, when there is no language?

Reply:
I can suggest
```kt
isApplicableForDefinedLanguage(language: String) = false

isApplicableForUndefinedLanguage(code:String) = false // covers the cases 1 and 3
```
I am not sure about the case 2. In the example with versions: 
V1 should know the difference from V2, and vice versa, V2 - from V1.
Thus, both versions should be in a single extension.
Otherwise, it makes an implementation of `isApplicable` more complicated and unobvious.

Reply:
Oh yeah, you're right. The second case doesn't need `isApplicable(code)` (== doesn't need to check the version/dialect), because it can just return `true` for `isApplicable(language)`, and then both language and code are already present in `buildCodeBlock(language?, code)`, so it can check the code here and render it respectively

two separate functions should be better in this case then :+1: